internal/daemon/api_logs.go

```diff
@@ -0,0 +1,318 @@
+	// finished. We can't just wait for the done channel here, because the
+	// client connection will close and iterateLogs may still be calling
+	// it.Next() when the caller calls it.Close(), causing a data race.
+	wg.Wait()
+}
+
+// Each log is written as a JSON object followed by a newline (JSON Lines):
+//
+// {"time":"2021-04-23T01:28:52.660Z","service":"redis","message":"redis started up"}
+// {"time":"2021-04-23T01:28:52.798Z","service":"thing","message":"did something"}
+type logEntry struct {
+	Time    time.Time `json:"time"`
+	Service string    `json:"service"`
+	Message string    `json:"message"`
+}
+
+type writeLogFunc func(log logEntry) error
+
+// Iterate through all logs in the given iterator, calling writeLog for each
+// until the iterator is exhausted.
+func iterateLogs(it servicelog.Iterator, br *bufio.Reader, cancel <-chan struct{}, now time.Time, writeLog writeLogFunc) error {
+	var prevTime time.Time
+	var prevService string
+	for it.Next(cancel) {
+		br.Reset(it)
+		for {
+			line, err := br.ReadSlice('\n')
+			if err != nil && !errors.Is(err, io.EOF) {
+				return err
+			}
+			if len(line) == 0 {
+				// No bytes read, we're done with this iteration
+				break
+			}
+
+			log, ok := parseLog(line)
+			if ok {
+				// Output normal log
+				err := writeLog(log)
+				if err != nil {
+					return err
+				}
+				if !now.IsZero() && log.Time.After(now) {
+					// Stop if we're past current time
+					return nil
+				}
+				prevTime = log.Time
+				prevService = log.Service
+			} else if !prevTime.IsZero() {
+				// Output partial log with previous timestamp and service name
+				err := writeLog(logEntry{prevTime, prevService, string(line)})
+				if err != nil {
+					return err
+				}
+			}
+
+			if errors.Is(err, io.EOF) {
+				// EOF reached, break after processing line
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// Parse a log entry of the form "2021-05-20T15:39:12.345Z [service] log message"
+func parseLog(line []byte) (logEntry, bool) {
+	fields := bytes.SplitN(line, []byte(" "), 3)
+	if len(fields) != 3 {
+		return logEntry{}, false
+	}
+	// .999 allows any number of fractional seconds (including none at all)
+	timestamp, err := time.Parse("2006-01-02T15:04:05.999Z07:00", string(fields[0]))
+	if err != nil {
+		return logEntry{}, false
+	}
+	if len(fields[1]) < 3 {
+		return logEntry{}, false
+	}
+	service := string(fields[1][1 : len(fields[1])-1]) // Trim [ and ] from "[service]"
+	message := string(fields[2])
+	return logEntry{timestamp, service, message}, true
+}
+
+// Writer that serializes concurrent writes to the underlying writer, and
+// calls Flush on the underlying writer after each write.
+type followWriter struct {
+	w     io.Writer
+	mutex *sync.Mutex
+}
+
+func (w *followWriter) Write(p []byte) (int, error) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	n, err := w.w.Write(p)
+	flushWriter(w.w) // Flush HTTP response after each line in follow mode
```

Comment:
Per earlier note, this function is not safe. Even if we ignore error/retry cases, Imagine what happens if any part of the stack writes half a JSON object because that's a perfectly fine way to use Write.

Reply:
Good call, thanks for pointing that out! I've done this by the sub-goroutines sending to a channel now, so there's only the main goroutine doing the writing.